multimodal/src/autogluon/multimodal/models/sam.py

```diff
@@ -0,0 +1,193 @@
+import logging
+import os
+import time
+import warnings
+from typing import Optional
+
+import torch
+import torch.nn.functional as F
+from omegaconf import DictConfig
+from torch import nn
+from transformers import SamConfig, SamModel
+
+from ..constants import COLUMN, IMAGE, IMAGE_VALID_NUM, LABEL, LOGITS, SEMANTIC_SEGMENTATION
+from .utils import assign_layer_ids, freeze_model_layers
+
+logger = logging.getLogger(__name__)
+
+
+class SAMForSemanticSegmentation(nn.Module):
+    """
+    Support SAM for binary real-world semantic segmentation.
+    Refer to https://huggingface.co/docs/transformers/main/model_doc/sam
+    """
+
+    def __init__(
+        self,
+        prefix: str,
+        checkpoint_name: str,
+        config: DictConfig,
+        pretrained: Optional[bool] = True,
+        frozen_layers: Optional[list] = None,
+    ):
+        """
+        Load a pretrained Segment Anything Model (SAM).
+
+        Parameters
+        ----------
+        prefix
+            The prefix of the SAMForSemanticSegmentation model.
+        checkpoint_name
+            Name of the SAM checkpoint.
+        pretrained
+            Whether using the pretrained SAM models. If pretrained=True, download the pretrained model.
+        frozen_layers
+            A list of substrings of frozen layers' names.
+        """
+
+        super().__init__()
+        self.prefix = prefix
+        self.pretrained = pretrained
+        self.checkpoint_name = checkpoint_name
+        self.frozen_layers = frozen_layers
+
+        self.device = None
+        self.name_to_id = {}
+
+        self._load_checkpoint(checkpoint_name)
+
+        freeze_model_layers(self.model, self.frozen_layers)
+
+        self.image_size = self.model.vision_encoder.image_size
+
+        # for Binary Semantic Segmentation Tasks.
+        self.model.mask_decoder.num_mask_tokens = 1
+        mask_token_data = self.model.mask_decoder.mask_tokens.weight.data[0]
+        self.model.mask_decoder.mask_tokens = nn.Embedding(1, self.model.mask_decoder.hidden_size)
+        self.model.mask_decoder.mask_tokens.weight.data[0] = mask_token_data
+        hyper_mlps = self.model.mask_decoder.output_hypernetworks_mlps[0]
+        self.model.mask_decoder.output_hypernetworks_mlps = nn.ModuleList([hyper_mlps])
+
+        self.config = self.model.config
+
+    def _load_checkpoint(self, checkpoint_name):
+        if self.pretrained:
+            self.model = SamModel.from_pretrained(checkpoint_name)
+        else:
+            configuration = SamConfig(name_or_path=checkpoint_name)
+            self.model = SamModel(configuration)
+
+    def set_data_preprocessor_device(self):
```

Comment:
Where do we need `set_data_preprocessor_device`? Lightning automatically handles the devices in training.